Replace proof Verification by go-ethereum impl

ed255 opened the pull request:
- Remove the custom Merkle Patricia Trie proof verification
  implementation
- Remove the custom RLP encoding/decoding implementation
- Use the Merkle Patricia Trie proof verification function from
  go-ethereum
  - This adds support to all types of non-existence proofs
- Use the RLP encoding/decoding functions from go-ethereum
- Add ~BytesHex~QuantityBytes and SliceData to abstract encoding of QUANTITY and slices of bytes in
  the StorageProof types

ed255 committed:
Replace proof Verification by go-ethereum impl

- Remove the custom Merkle Patricia Trie proof verification
  implementation
- Remove the custom RLP encoding/decoding implementation
- Use the Merkle Patricia Trie proof verification function from
  go-ethereum
  - This adds support to all types of non-existence proofs
- Use the RLP encoding/decoding functions from go-ethereum
- Add BytesHex and SliceBytesHex to abstract encoding of byte
  slices in the StorageProof types

p4u commented on ethstorageproof/types.go:
```diff
@@ -1,25 +1,126 @@
+	dec := make([]byte, len(input)/2)
+	if _, err := hex.Decode(dec, input); err != nil {
+		return err
+	} else {
```
No need for this else

p4u commented on ethstorageproof/types.go:
```diff
@@ -1,25 +1,126 @@
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (b *BytesHex) UnmarshalText(input []byte) error {
+	if bytes.HasPrefix(input, []byte("0x")) {
```
You can use `bytes.TrimPrefix()` instead

p4u commented on ethstorageproof/types.go:
```diff
@@ -1,25 +1,126 @@
+	value, ok := m.kvs[h]
+	if ok {
+		return value, nil
+	} else {
```
No need for this else

p4u commented on ethstorageproof/types.go:
```diff
@@ -1,25 +1,126 @@
+	if ok {
+		return value, nil
+	} else {
+		return nil, nil
```
Actually I think this logic should be reversed and an error should be returned if key is not found.

```
if !ok { return nil, fmt.Errorf("...") }
return value, nil
```

p4u commented on ethstorageproof/types.go:
```diff
@@ -1,25 +1,126 @@
+// Put sets or updates the value at key
+func (m *MemDB) Put(key []byte, value []byte) error {
+	var h common.Hash
+	copy(h[:], key)
```
You can use `common.BytesToHash()` instead

p4u commented:
It would be nice to add some extra test to this "new code". We could use the Proofs hardcoded on this test from vocdoni-node
https://github.com/vocdoni/vocdoni-node/blob/master/vochain/proof_test.go#L574

What do you think @ed255?

ed255 committed:
Add more proofs to test

Add proofs from
https://github.com/vocdoni/vocdoni-node/blob/5b292fc7e74327f0127908636072482660b060b2/vochain/proof_test.go#L574

Add a minime storage proof of non-existence that couldn't be verified
with the previous VerifyProof implementation.

ed255 commented:
> It would be nice to add some extra test to this "new code". We could use the Proofs hardcoded on this test from vocdoni-node
> https://github.com/vocdoni/vocdoni-node/blob/master/vochain/proof_test.go#L574
> 
> What do you think @ed255?

I've added a new commit which will:
- Add proofs from https://github.com/vocdoni/vocdoni-node/blob/5b292fc7e74327f0127908636072482660b060b2/vochain/proof_test.go#L574
- Add a minime storage proof of non-existence that couldn't be verified with the previous VerifyProof implementation.

I had to edit the `value` fields from https://github.com/vocdoni/vocdoni-node/blob/5b292fc7e74327f0127908636072482660b060b2/vochain/proof_test.go#L574 because they were even-length hex strings (I added a padding 0).  How where those proofs generated?

p4u commented:
RLP encoding can produce even-length hex strings, i.e 0x0 means "null" for RLP. So the Marshaling/Unmarshaling should take this into account, instead of using hex.Decode() you probably want to use some specific RLP function. Or you can just list the cases where the length can be even and modify the original hex string accordingly. 

I.e https://github.com/vocdoni/storage-proofs-eth-go/blob/master/ethstorageproof/ethstorageproof.go#L339

p4u commented:
Whenever this is addresses, ping me please :)

ed255 committed:
Replace BytesHex in StorageProof by QUANTITY

p4u commented on ethstorageproof/ethstorageproof_test.go:
```diff
@@ -153,8 +154,18 @@ func TestVerify(t *testing.T) {
+		if len(tt.value) != 0 {
+			value, err = rlp.EncodeToBytes(tt.value)
+			if err != nil {
+				panic(err)
```
We usually use t.Fatal() instead of panic()

p4u replied:
Actually we are slowly migrating to quicktest

`import qt "github.com/frankban/quicktest"`

`qt.Assert(b, err, qt.IsNil)`

But if you don't want to change it on this PR we could do it in the future.

ed255 replied:
I've added an issue to migrate to quicktest: https://github.com/vocdoni/storage-proofs-eth-go/issues/5

p4u commented on ethstorageproof/ethstorageproof_test.go:
```diff
@@ -172,16 +183,12 @@ func toBytes(s string) []byte {
+func asQuantity(s string) []byte {
+	var q QuantityBytes
+	if err := q.UnmarshalText([]byte(s)); err != nil {
 		panic(err)
```
same

p4u commented on ethstorageproof/types.go:
```diff
@@ -11,45 +10,42 @@ import (
-	hex.Encode(result[2:], b)
-	return result, nil
+func (q QuantityBytes) MarshalText() ([]byte, error) {
+	return []byte(fmt.Sprintf("0x%#x", new(big.Int).SetBytes(q))), nil
```
Can you add a comment on this expression? `0x%#x`

ed255 replied:
I've replaced this expression to be more explicit about what's going on.  Actually there was a but, it shouldn't have bee `0x%#x` but `%#x`.  What `%#x` did was format the big.Int in hexadecimal, taking it as an integer (not as a byte slice).  I think it's more clear now, where I use `big.Int.Text(16)`.

p4u commented:
LGTM now, just minor comments. You can address them or let it for a future PR.

Please, "rebase and merge" yourself.
